design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,542 @@
+- **DO** take ownership of the xlang error info using attach semantics.
+
+## Translation of errors
+
+xlang's goal is to allow a component to be written using natural and familiar code for the language
+it is implemented in and for it to be consumed using natural and familiar code for the language which
+the consumer is written in. To achieve this, xlang needs to translate errors from the natural form
+of one language to a natural form of that error in another language.
+
+But different languages have different errors which can occur and not all errors in one language
+always directly map to errors from another language. In addition, maintaining a mapping of errors
+between each supported language is not scalable. Lastly, there are a lot of errors which are domain
+specific and may not make sense in other languages and thereby not have equivalent error codes natively
+available. Because of this, xlang will only support a limited set of errors which can be used by the
+projection of a language to map to or from its error. These errors are intended to be used to
+indicate an error during the translation of the call and not to indicate errors from components.
+Components are encouraged to avoid returning errors and use the Try Pattern as discussed later in
+[Alternative to errors](#alternative-to-errors).
+
+If the error being mapped has an equivalent in the set of errors and the projection of the consumer's
+language provides an equivalent mapping from that mapped error, then that error will have a natural
+mapping in the other language. In the case there isn't an equivalent error to map to in xlang, then a
+generic error, `xlang_fail`, will be available to be used. Note that in this case, the natural mapping of
+errors from one language to the other will be lost in the sense that it will not be an equivalent mapping.
+In the case there isn't an equivalent error to map to in a consumer's language, it is left to the projection
+of that language to decide what error from its language to map to based on what makes sense for that language.
+This can end up being a generic error.
+
+## xlang errors
+
+xlang will use an `Int32`, `xlang_result`, to represent its errors. The representation of this 32-bit
+value will be based on the same model used in COM for HRESULTs to assist with sharing error codes,
+but will not necessarily be defined in this design note on a bit by bit basis as is in COM.
+This is because xlang will not inherit all the HRESULT errors and will instead define a set of
+errors with their values defined. This will allow xlang to avoid defining all the possible HRESULT
+errors on each supported platform especially when most will not be used as xlang will be scope its errors
+to ones that are likely to occur during translation of calls. This will also allow for projections to
+avoid needing to maintain a large mapping of errors.
+
+The following errors will be defined in xlang:
+
+xlang error          | Value      | Description
```

Comment:
per arch discussion today, agreed that a zero-based monotonically increasing enum has a few advantages:
1. it does not confuse cross plat work by injecting Windows-isms
2. it permits direct indexing into a table (e.g., of exception factory methods)
3. it enables a simple version check for projections to translate errors they know of (<xlang_error_count) and handling all others (>=xlang_error_count) 'generically'

Reply:
Addressed.